Enable PermutationFeatureImportance for Regression

shmoradims opened the pull request:
This is a PR to enable PermutationFeatureImportance feature for model explainability. It shows how to do PFI for a regression model.

Discussion points / open issues:
* Namespace: I'm having some circular dependency issues, so MLContext.Regression.PermutationFeatureImportance() is not fully implemented yet. I've put a comment in there.
* Final name: PermutationFeatureImportance or something else?
* Return type: It used to be IDataView in TLC. I suggest we return List<(string featureName, RegressionEvaluator.Result metricsDelta)> instead.
* Weight Filtering: TLC had weight filtering features by using IPredictorWithFeatureWeights. We can't do this in ML.NET so I removed that option.
* Report file: TLC produces a text report file. I left that option out.

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceRegression.cs:
```diff
@@ -0,0 +1,233 @@
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Float = System.Single;
```
Can you not bring it? Just use `float`

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceRegression.cs:
```diff
@@ -0,0 +1,233 @@
+                data.Schema.TryGetColumnIndex(features, out int featuresColumnIndex);
+
+                // Todo: check this
+                //TrainerUtils.CheckFeatureFloatVector(roleMapped);
```
is it something you plan to uncomment, or it can be deleted?

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceRegression.cs:
```diff
@@ -0,0 +1,233 @@
+            string label = DefaultColumnNames.Label, string features = DefaultColumnNames.Features,
+            int topExamples = 0, int progressIterations = 10)
+        {
+            var host = ((IHostEnvironment)_mlContext).Register(nameof(GetImportanceMetricsMatrix));
```
>var host = ((IHostEnvironment)_mlContext).Register(nameof(GetImportanceMetricsMatrix)); [](start = 12, length = 87)

Can you move host creation to constructor of class rather than it's method?

shmoradims commented:
> This is a WIP-PR to enable PermutationFeatureImportance feature for model explainability. It shows how to do PFI for a regression model.
> 
> Discussion points / open issues:
> 
> * Namespace: I'm having some circular dependency issues, so MLContext.Regression.PermutationFeatureImportance() is not fully implemented yet. I've put a comment in there.
> * Final name: PermutationFeatureImportance or something else?
> * Return type: It used to be IDataView in TLC. I suggest we return List<(string featureName, RegressionEvaluator.Result metricsDelta)> instead.
> * Weight Filtering: TLC had weight filtering features by using IPredictorWithFeatureWeights. We can't do this in ML.NET so I removed that option.
> * Report file: TLC produces a text report file. I left that option out.

Updates after talking with Pete:

* Changed return type to ImmutableArray<RegressionEvaluator.Result>
* Fixed the namespace issue

yaeldekel commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,224 @@
+                data.Schema.TryGetColumnIndex(features, out int featuresColumnIndex);
+
+                // Todo: check this
+                //TrainerUtils.CheckFeatureFloatVector(roleMapped);
```
This can be deleted.

yaeldekel commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,224 @@
+{
+    using Stopwatch = System.Diagnostics.Stopwatch;
+
+    internal sealed class PermutationFeatureImportance<TResult>
```
Are there any plans to introduce the PFI command here?

shmoradims replied:
The priority is to get pfi in 0.8, which doesn't need the command. So pfi command would come later.

---
In reply to: [234781545](https://github.com/dotnet/machinelearning/pull/1658#discussion_r234781545) [](ancestors = 234781545)

Zruty0 commented on src/Microsoft.ML.Data/TrainContext.cs:
```diff
@@ -21,6 +21,8 @@ namespace Microsoft.ML
     {
         protected readonly IHost Host;
+
+        [BestFriend]
```
>BestFriend [](start = 9, length = 10)

This should be reverted. Use `GetEnvironment` extension method

Zruty0 commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,224 @@
+
+    internal sealed class PermutationFeatureImportance<TResult>
+    {
+        private readonly IHostEnvironment _host;
```
>_host [](start = 42, length = 5)

`_env`

Zruty0 commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,224 @@
+{
+    using Stopwatch = System.Diagnostics.Stopwatch;
+
+    internal sealed class PermutationFeatureImportance<TResult>
```
>sealed [](start = 13, length = 6)

Make this a static class. There is no need to have an instance, and also you have a bug when you call the PFI method more than once.

Zruty0 commented:
> using Microsoft.ML.Core.Data;

Header comments for new files

---
Refers to: src/Microsoft.ML.Transforms/PermutationFeatureImportanceExtensions.cs:1 in 64ebc6f. [](commit_id = 64ebc6feec1497f5a443cd14ed543c9adc85739c, deletion_comment = False)

Zruty0 commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+    public static class PermutationFeatureImportanceExtensions
+    {
+        public static ImmutableArray<RegressionEvaluator.Result>
+            PermutationFeatureImportance(
```
>PermutationFeatureImportance [](start = 12, length = 28)

`CalculateFeatureImportance` (to make it a verb). 
Also add a summary comment and describe all the parameters

Zruty0 commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,224 @@
+                Func<TResult, TResult, TResult> deltaFunc,
+                string features,
+                int topExamples,
+                int progressIterations)
```
>progressIterations [](start = 20, length = 18)

There is no point in having this as a top-level parameter. Just print progress every 10th feature, that's all. Or every feature.

Zruty0 commented on src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs:
```diff
@@ -212,6 +212,16 @@ internal Result(IExceptionContext ectx, IRow overallResult)
             }
+
+            [BestFriend]
+            internal Result(double l1, double l2, double rms, double lossfn, double rsquared)
```
camelCase please: `lossFunction`, `rSquared`

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,212 @@
+                        ch.Info(string.Format("Processed slots {0} - {1} in {2}", processedCnt - progressIterations,
+                            processedCnt, stopwatch.Elapsed));
+                        stopwatch.Restart();
+                    }
```
We have progress channel for this.
Please don't use stopwatch.

[https://github.com/dotnet/machinelearning/blob/0fe8bafa11e0b759425100db12327569a8c444bb/src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs#L344](https://github.com/dotnet/machinelearning/blob/0fe8bafa11e0b759425100db12327569a8c444bb/src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs#L344)
here is example of how it should be done.

just set header with one metric: processed slots, use SetProgress without limit, and set checkpoint after your loop.

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,212 @@
+                    while (initialfeatureValuesList.Count < maxSize && cursor.MoveNext())
+                    {
+                        featuresGetter(ref featuresBuffer);
+                        initialfeatureValuesList.Add(featuresBuffer.GetItemOrDefault(workingFeatureIndices[0]));
```
>workingFeatureIndices[0] [](start = 85, length = 24)

this looks suspicious.
`workingFeatureIndices[0]` is always 0, why we need `workingFeatureIndices[0]`

Ivanidzo4ka replied:
Honestly I don't understand why we even need `workingFeatureIndices` glancing through code, we never modify it, and it just array with values from 0 to numSlots.

---
In reply to: [235190185](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235190185) [](ancestors = 235190185)

Ivanidzo4ka replied:
OK, I've looked into internal repo, and we have one specific case for the model with weights and we populate that array differently for that case.
Why it got dropped?

---
In reply to: [235190701](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235190701) [](ancestors = 235190701,235190185)

ganik replied:
Here is the comment from Shahab: 
* Weight Filtering: TLC had weight filtering features by using IPredictorWithFeatureWeights. We can't do this in ML.NET so I removed that option.


---
In reply to: [235195353](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235195353) [](ancestors = 235195353,235190701,235190185)

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+        /// <param name="data">The evaluation data set.</param>
+        /// <param name="label">Label column name.</param>
+        /// <param name="features">Feature column names.</param>
+        /// <param name="topExamples">Number of top (highest positive) contributions to output.</param>
```
>Number of top (highest positive) contributions to output. [](start = 38, length = 57)

What?
This is how many examples you want to take from evaluation data set. What that highest positive contributions even mean?

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportanceExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+                IDataView data,
+                string label = DefaultColumnNames.Label,
+                string features = DefaultColumnNames.Features,
+                int topExamples = 0)
```
>int topExamples = 0 [](start = 16, length = 19)

I would rather make it nullable by default, 0 is really counterintuitive. And put check on positivity. Either provide us how many examples you want (positive number) or set null and we will use whole dataset.

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,212 @@
+            var host = env.Register(nameof(PermutationFeatureImportance<TResult>));
+            host.CheckValue(model, nameof(model));
+            host.CheckValue(data, nameof(data));
+            host.CheckValue(features, nameof(features));
```
>CheckValue [](start = 17, length = 10)

CheckNonEmpty or CheckNonWhiteSpace. 
This is a string CheckValue would be happy to take "" but this would be invalid feature column name.

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,281 @@
+                        nextValues.CopyTo(featureValuesBuffer, 0);
+                        Array.Clear(nextValues, 0, nextValues.Length);
+
+                        pch.Checkpoint(processedCnt, processedCnt);
```
>processedCnt [](start = 39, length = 12)

Checkpoint should be outside of loop.

rogancarr replied:
Done

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,281 @@
+    internal static class PermutationFeatureImportance<TResult>
+    {
+        public static ImmutableArray<TResult>
+            GetImportanceMetricsMatrix(
```
If you have time, can you refactor this method into at least two chunks, initialization and  for loop on top of permutation routine. Or maybe just put permutation routine into separate function.
In current state it's quite hard to read. (~200 lines)

Ivanidzo4ka commented on test/Microsoft.ML.Tests/PermutationFeatureImportanceTests.cs:
```diff
@@ -0,0 +1,200 @@
+// <copyright company="Microsoft Corporation">
+//     Copyright (c) Microsoft Corporation. All rights reserved.
+// </copyright>
+//------------------------------------------------------------------------------
```
wrong header.

rogancarr replied:
Done.

Ivanidzo4ka commented on test/Microsoft.ML.Tests/PermutationFeatureImportanceTests.cs:
```diff
@@ -0,0 +1,200 @@
+// </copyright>
+//------------------------------------------------------------------------------
+
+using Float = System.Single;
```
please just use `float` in code, and delete this one.

rogancarr replied:
done.

Ivanidzo4ka commented on test/Microsoft.ML.Tests/PermutationFeatureImportanceTests.cs:
```diff
@@ -0,0 +1,200 @@
+using Xunit;
+using Xunit.Abstractions;
+using Microsoft.ML.Transforms;
+using System.Collections.Immutable;
```
would be cool to sort this usings

rogancarr replied:
Done

rogancarr commented on src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs:
```diff
@@ -0,0 +1,281 @@
+
+                // Now iterate through all the working slots, do permutation and calc the delta of metrics.
+                int processedCnt = 0;
+                int j = 0;
```
Done.
